vnext/ReactUWP/Threading/UIMessageQueueThread.cpp

```diff
@@ -17,25 +17,41 @@ UIMessageQueueThread::UIMessageQueueThread(
 UIMessageQueueThread::~UIMessageQueueThread() {}
 
 void UIMessageQueueThread::runOnQueue(std::function<void()> &&func) {
-  m_uiDispatcher.RunAsync(
-      winrt::Windows::UI::Core::CoreDispatcherPriority::Normal,
-      [func = std::move(func)]() {
+  ensureQueue();
```

Comment:
there seems to be a possible thread safety issue here, we need a lock for runOnQueue and onBatchComplete to guard access to m_queue?  Usually everything is queued and batch complete is done on the same thread, but since nativemodules/3rd party code also can add to this queue and do it from any thread, we should be safe or add some check ensuring that you can only queue from an expected thread

Reply:
As a starting point I'll put a stake in the ground that this queue should only be used for communication between the JS thread and the UI thread.  Added a thread check in DEBUG builds.

---
In reply to: [314421719](https://github.com/microsoft/react-native-windows/pull/2937#discussion_r314421719) [](ancestors = 314421719)